### DISCUSSION TITLE
How to republish to Confluence?

### DISCUSSION THREAD

**Discussion by obfischer** [docToolchain/docToolchain#1068] [Q&A]
Dear all,

I managed to publish my documentation with docToolchain to our confluence and the result is very nice. But I can't publish the same configuration twice, I always have to delete the previously published page tree.

How can I simply update the existing pages?

**Comment by rdmueller** [comment]
thanx for bringing up this question.

`publishToConfluence` generates a hash of every single page and stores it along with the page in confluence (should be hidden at the end of the page).
As soon as you try to re-publish, the local hash and remote hash are compared and the page will not be republished if the hashes are the same, since the content will be the same.
So basically, a republish should work if you change the content of your docs.

But I guess, from how you ask the question, this mechanism doesn't work for you, does it? Which dtc version do you use? local install or docker?

**Reply by obfischer** [reply]
Hi @rdmueller, the mechanism works and it is totally fine, how it works. But it would b nice not to send with an error message. A message like: _Skipping page..._ would nicer.

If you can send me a bill, I would order this feature. ;-)

P.S.: I try to publish our documentation during every build in our CI pipeline.

**Reply by rdmueller** [reply]
Oh, this change should be a no brainer!
What is the current output?

**Reply by obfischer** [reply]
I have to check it. I am currently not on my desk. More important than the output is the exit code.

**Reply by rdmueller** [reply]
I think we already have a fix for the exit code in the next release, but I have to check this.

**Reply by obfischer** [reply]
Hi @rdmueller,

I am still on this issue. Simply changing the content of a Asciidoc files does not allow to republish the documentation. Is it because I am using v2.2.0? I am afraid to fail with my documentation as code approach, as I must provide the documentation via Confluence.

**Comment by obfischer** [comment]
Hi @rdmueller,

here is the output.

```
publish /Users/plx/temp/beispiel/build/html5/arc42/arc42.html
Template
Start getting headers
.

> Task :publishToConfluence FAILED

FAILURE: Build failed with an exception.

* Where:
Script '/Users/plx/.doctoolchain/docToolchain-2.2.0/scripts/publishToConfluence.gradle' line: 30

* What went wrong:
Execution failed for task ':publishToConfluence'.
> Cannot create page, page with the same title=aTemplate with id=163471595 already exists in the space. A Confluence page title must be unique within a space, consider specifying a 'confluencePagePrefix' in ConfluenceConfig.groovy

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 38s
2 actionable tasks: 1 executed, 1 up-to-date
```

Exitcode is 1.

**Reply by rdmueller** [reply]
oh, this is a problem from which the script can't recover. I am not sure about the problem, so let me explain the situation:

confluence does not allow to have two pages with the same name in one space. 
For each page, the script searches for an already existing page and checks if is below the configured ancestorId - the page root for the doctoolchain pages. If it is not, docToolchain will not overwrite it, because it does not belong to the current import. It will throw the error above instead.

In the past, we had sometimes problems with the confluence search api. Some special characters within the page title resulted in wrong search results. But this problem should be solved.

Have you checked that the configured `ancestorName` or `ancestorId` exists?

**Reply by obfischer** [reply]
I am only able to publish to Confluence if `ancestorId` is not set. See issue #1074 for details. 

As Confluence is the only plattform I can use to publish the documentation, I have to discard the current approach and to work with Confluence in the traditionl way 😢

**Reply by rdmueller** [reply]
please try v2.2.1

**Reply by obfischer** [reply]
With v2.2.1 I am able to publish the docs as often as I want. Now only the presence of `ancestorId` leads to the issue described in #1074.
